File: backend/app/api/routes/employees.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import Optional

from app.core.database import get_db
from app.schemas.employee import (
    EmployeeCreate,
    EmployeeUpdate,
    EmployeeResponse,
    PaginatedEmployeeResponse,
)
from app.services import employee_service

router = APIRouter(prefix="/employees", tags=["employees"])


@router.get("", response_model=PaginatedEmployeeResponse)
def list_employees(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    search: Optional[str] = Query(default=None),
    country: Optional[str] = Query(default=None),
    department: Optional[str] = Query(default=None),
    job_title: Optional[str] = Query(default=None),
    db: Session = Depends(get_db),
):
    return employee_service.get_employees(
        db, page, page_size, search, country, department, job_title
    )


@router.post("", response_model=EmployeeResponse, status_code=201)
def create_employee(data: EmployeeCreate, db: Session = Depends(get_db)):
    return employee_service.create_employee(db, data)


@router.get("/{employee_id}", response_model=EmployeeResponse)
def get_employee(employee_id: int, db: Session = Depends(get_db)):
    employee = employee_service.get_employee_by_id(db, employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    return employee


@router.put("/{employee_id}", response_model=EmployeeResponse)
def update_employee(
    employee_id: int,
    data: EmployeeUpdate,
    db: Session = Depends(get_db),
):
    employee = employee_service.update_employee(db, employee_id, data)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    return employee


@router.delete("/{employee_id}", status_code=204)
def delete_employee(employee_id: int, db: Session = Depends(get_db)):
    deleted = employee_service.delete_employee(db, employee_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Employee not found")


@router.get("/meta/countries", response_model=list[str])
def list_countries(db: Session = Depends(get_db)):
    """Returns distinct countries — used to populate filter dropdowns in the UI."""
    return employee_service.get_distinct_countries(db)


@router.get("/meta/departments", response_model=list[str])
def list_departments(db: Session = Depends(get_db)):
    return employee_service.get_distinct_departments(db)


@router.get("/meta/job-titles", response_model=list[str])
def list_job_titles(db: Session = Depends(get_db)):
    return employee_service.get_distinct_job_titles(db)

File: frontend/src/api/employees.ts
import { apiClient } from "./client";

export interface Employee {
  id: number;
  full_name: string;
  email: string;
  job_title: string;
  department: string;
  country: string;
  salary: number;
  currency: string;
  employment_type: string;
  hire_date: string;
  created_at: string;
  updated_at: string;
}

export interface EmployeeCreate {
  full_name: string;
  email: string;
  job_title: string;
  department: string;
  country: string;
  salary: number;
  currency: string;
  employment_type: string;
  hire_date: string;
}

export type EmployeeUpdate = Partial<EmployeeCreate>;

export interface PaginatedResponse {
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  data: Employee[];
}

export interface EmployeeFilters {
  page?: number;
  page_size?: number;
  search?: string;
  country?: string;
  department?: string;
  job_title?: string;
}

// ── CRUD ──────────────────────────────────────────────────────────────────────

export const getEmployees = async (filters: EmployeeFilters = {}) => {
  const { data } = await apiClient.get<PaginatedResponse>("/employees", {
    params: filters,
  });
  return data;
};

export const getEmployee = async (id: number) => {
  const { data } = await apiClient.get<Employee>(`/employees/${id}`);
  return data;
};

export const createEmployee = async (payload: EmployeeCreate) => {
  const { data } = await apiClient.post<Employee>("/employees", payload);
  return data;
};

export const updateEmployee = async (id: number, payload: EmployeeUpdate) => {
  const { data } = await apiClient.put<Employee>(`/employees/${id}`, payload);
  return data;
};

export const deleteEmployee = async (id: number) => {
  await apiClient.delete(`/employees/${id}`);
};

// ── Meta (for dropdowns) ──────────────────────────────────────────────────────

export const getCountries = async () => {
  const { data } = await apiClient.get<string[]>("/employees/meta/countries");
  return data;
};

export const getDepartments = async () => {
  const { data } = await apiClient.get<string[]>("/employees/meta/departments");
  return data;
};

export const getJobTitles = async () => {
  const { data } = await apiClient.get<string[]>("/employees/meta/job-titles");
  return data;
};